src/Native/System.Native/pal_networking.cpp

```diff
@@ -287,15 +299,211 @@ extern "C" int32_t GetHostEntryForName(const uint8_t* address, HostEntry* entry)
+{
+    if (hostname == nullptr || entry == nullptr)
+    {
+        return PAL_BAD_ARG;
+    }
+
+    hostent* hostEntry = nullptr;
+    int error = 0;
+
+#if defined(HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR)
+    hostEntry = gethostbyname(reinterpret_cast<const char*>(hostname));
+    error = h_errno;
+#elif HAVE_GETHOSTBYNAME_R
+    error = GetHostByNameHelper(hostname, &hostEntry);
+#else
+#error Platform does not provide thread-safe gethostbyname
+#endif
+
+    if (hostEntry == nullptr)
+    {
+        return ConvertGetHostErrorPlatformToPal(error);
+    }
+
+    ConvertHostEntPlatformToPal(*entry, *hostEntry);
+    return PAL_SUCCESS;
+}
+
+#if HAVE_GETHOSTBYADDR_R
+static int GetHostByAddrHelper(const uint8_t* addr, const socklen_t addrLen, int type, hostent** entry)
+{
+    assert(addr != nullptr);
+    assert(addrLen >= 0);
+    assert(entry != nullptr);
+
+    size_t scratchLen = 512;
+
+    for (;;)
+    {
+        uint8_t* buffer = reinterpret_cast<uint8_t*>(malloc(sizeof(hostent) + scratchLen));
+        if (buffer == nullptr)
+        {
+            return PAL_NO_MEM;
+        }
+
+        hostent* result = reinterpret_cast<hostent*>(buffer);
+        char* scratch = reinterpret_cast<char*>(&buffer[sizeof(hostent)]);
+
+        int getHostErrno;
+        int err = gethostbyaddr_r(addr, addrLen, type, result, scratch, scratchLen, entry, &getHostErrno);
+        switch (err)
+        {
+            case 0:
+                *entry = result;
+                return 0;
+
+            case ERANGE:
+                free(buffer);
+                scratchLen *= 2;
+                break;
+
+            default:
+                free(buffer);
+                *entry = nullptr;
+                return getHostErrno;
+        }
+    }
+}
+#endif
+
+extern "C" int32_t GetHostByAddress(const IPAddress* address, HostEntry* entry)
+{
+    if (address == nullptr || entry == nullptr)
+    {
+        return PAL_BAD_ARG;
+    }
+
+    uint8_t* addr = nullptr;
+    socklen_t addrLen = 0;
+    int type = AF_UNSPEC;
+
+    in_addr inAddr = {};
+    in6_addr in6Addr = {};
+
+    if (!address->IsIPv6)
+    {
+        ConvertByteArrayToInAddr(inAddr, address->Address, NUM_BYTES_IN_IPV4_ADDRESS);
+        addr = reinterpret_cast<uint8_t*>(&inAddr);
+        addrLen = sizeof(inAddr);
+        type = AF_INET;
+    }
+    else
+    {
+        ConvertByteArrayToIn6Addr(in6Addr, address->Address, NUM_BYTES_IN_IPV6_ADDRESS);
+        addr = reinterpret_cast<uint8_t*>(&in6Addr);
+        addrLen = sizeof(in6Addr);
+        type = AF_INET6;
+    }
+
+    hostent* hostEntry = nullptr;
+    int error = 0;
+
+#if defined(HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR)
+    hostEntry = gethostbyaddr(addr, addrLen, type);
+    error = h_errno;
+#elif HAVE_GETHOSTBYADDR_R
+    error = GetHostByAddrHelper(addr, addrLen, type, &hostEntry);
+#else
+#error Platform does not provide thread-safe gethostbyname
```

Comment:
> I think we're going to hit this on FreeBSD.

FreeBSD also provides thread-safe implementations of gethostby{name,addr}.

> Everything I can find suggests that getaddrinfo and getnameinfo are considered replacements for gethostbyname and gethostybyaddr and it seems to me they could be used to get portable thread-safety. Can we use them everywhere...?

We're deliberately using `gethostby{name,addr}` in certain situations in order to match the behavior of the Windows implementation (which also uses these functions). It might be possible to switch these functions to `get{addr,name}info`, but it would require some investigation (e.g. I'm not sure if gai/gni has any equivalent for `h_aliases`).

FWIW, supporting platforms without thread-safe `gethostby{name,addr}` is straightforward: making a deep copy of the returned data is all that's necessary.